azure-pipelines.yml

```diff
@@ -0,0 +1,343 @@
+        PYTHON_VERSION: '2.7'
+        PYSMT_SOLVER: "msat_wrap"
+      Py27Z3Wrap:
+        python.version: '2.7'
+        PYTHON_VERSION: '2.7'
+        PYSMT_SOLVER: "z3_wrap"
+
+  steps:
+  - task: UsePythonVersion@0
+    inputs:
+      versionSpec: '$(python.version)'
+      architecture: 'x64'
+  - bash: ci/install_linux.sh
+    displayName: 'Install and Test'
+
+
+
+- job: 'Ubuntu16PyPy2'
+  condition: or(eq(variables['Build.SourceBranchName'], 'master'), eq(variables['Build.Reason'], 'PullRequest'))
+  pool:
+    vmImage: 'Ubuntu-16.04'
+  strategy:
+    matrix:
+      PyPy2NoSolver:
+        pypy_binary: 'pypy'
+        PYTHON_VERSION: 'pypy'
+        PYSMT_SOLVER: "None"
+      PyPy2MsatWrap:
+        pypy_binary: 'pypy'
+        PYTHON_VERSION: 'pypy'
+        PYSMT_SOLVER: "msat_wrap"
+      PyPy2Z3Wrap:
+        pypy_binary: 'pypy'
+        PYTHON_VERSION: 'pypy'
+        PYSMT_SOLVER: "z3_wrap"
+
+  steps:
+  - bash: |
+      echo "##vso[task.setvariable variable=PATH]$(System.DefaultWorkingDirectory)/pypy.tmp/pypy/bin:$(PATH)"
+  - bash: |
+      mkdir pypy.tmp
+      cd pypy.tmp
+      curl --location --output archive https://bitbucket.org/pypy/pypy/downloads/pypy2.7-v7.0.0-linux64.tar.bz2
+      tar xvf archive
+      rm archive
+      mv ./* pypy
+      cd ..
+    displayName: 'Get PyPy'
+  - bash: |
+      type $(pypy_binary)
+      $(pypy_binary) --version
+    displayName: 'Check for PyPy'
+  - bash: |
+      curl --output pypy.tmp/get-pip.py https://bootstrap.pypa.io/get-pip.py
+      $(pypy_binary) pypy.tmp/get-pip.py
+    displayName: 'Get PIP'
+  - bash: ci/install_linux.sh
+    displayName: 'Install and Test'
+
+
+
+- job: 'Ubuntu16PyPy3'
+  condition: or(eq(variables['Build.SourceBranchName'], 'master'), eq(variables['Build.Reason'], 'PullRequest'))
+  pool:
+    vmImage: 'Ubuntu-16.04'
+  strategy:
+    matrix:
+      PyPy3NoSolver:
+        pypy_binary: 'pypy3'
+        PYTHON_VERSION: 'pypy3'
+        PYSMT_SOLVER: "None"
+      PyPy3MsatWrap:
+        pypy_binary: 'pypy3'
+        PYTHON_VERSION: 'pypy3'
+        PYSMT_SOLVER: "msat_wrap"
+      PyPy3Z3Wrap:
+        pypy_binary: 'pypy3'
+        PYTHON_VERSION: 'pypy3'
+        PYSMT_SOLVER: "z3_wrap"
+
+  steps:
+  - bash: |
+      echo "##vso[task.setvariable variable=PATH]$(System.DefaultWorkingDirectory)/pypy.tmp/pypy/bin:$(PATH)"
+  - bash: |
+      mkdir pypy.tmp
+      cd pypy.tmp
+      curl --location --output archive https://bitbucket.org/pypy/pypy/downloads/pypy3.5-v7.0.0-linux64.tar.bz2
+      tar xvf archive
+      rm archive
+      mv ./* pypy
+      cd ..
+    displayName: 'Get PyPy'
+  - bash: |
+      type $(pypy_binary)
+      $(pypy_binary) --version
+    displayName: 'Check for PyPy'
+  - bash: |
+      curl --output pypy.tmp/get-pip.py https://bootstrap.pypa.io/get-pip.py
+      $(pypy_binary) pypy.tmp/get-pip.py
+    displayName: 'Get PIP'
+  - bash: ci/install_linux.sh
+    displayName: 'Install and Test'
+
+
+
+- job: 'Ubuntu16Py37'
```

Comment:
I would run this as first job, then the Ubuntu16All and then the Ubutnu16Py27. Failures in this one give us a better localization of the problems.